src/js_realm.hpp

```diff
@@ -213,6 +213,68 @@ class RealmDelegate : public BindingContext {
     friend class RealmClass<T>;
 };
 
+
+template<typename T>
+class ShouldCompactOnLaunchFunctor {
+public:
+    ShouldCompactOnLaunchFunctor(typename T::Context ctx, typename T::Function should_compact_on_launch_func)
+    : m_ctx(Context<T>::get_global_context(ctx))
+    , m_func(ctx, should_compact_on_launch_func)
+    , m_event_loop_dispatcher {ShouldCompactOnLaunchFunctor<T>::main_loop_handler}
+    , m_mutex{new std::mutex}
```

Comment:
Could remove the shared pointer in this context. 
It means only main is responsible for destructing the mutex which is fine since main will wait for the result first and only then will complete

Reply:
So just declare these members on the class and they will be initialized for you

Reply:
these are required actually so the Functor can be moved.